src/errors.rs

```diff
@@ -171,57 +171,69 @@ fn remove_component_msg(cs: &Component, manifest: &Manifest, toolchain: &str) ->
+                );
+            }
+             _ = write!(
+                buf,
+                "If you are trying to update an existing toolchain and no longer need this component,\n\
+                you might want to remove it first with:\n\n{}\n\n{TOOLSTATE_MSG}",
+                remove_component_msg(c, manifest, toolchain)
             );
         }
-        let _ = write!(
-            buf,
-            "If you don't need the component, you can remove it with:\n\n{}",
-            remove_component_msg(&cs[0], manifest, toolchain)
-        );
-    } else {
-        // More than one component
+        cs => {
+            // More than one component
+            let same_target = cs
+                .iter()
+                .all(|c| c.target == cs[0].target || c.target.is_none());
 
-        let same_target = cs
-            .iter()
-            .all(|c| c.target == cs[0].target || c.target.is_none());
+            let cs_str = if same_target {
+                cs.iter()
+                    .map(|c| format!("'{}'", c.short_name(manifest)))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            } else {
+                cs.iter()
+                    .map(|c| c.description(manifest))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            };
 
-        let cs_str = if same_target {
-            cs.iter()
-                .map(|c| format!("'{}'", c.short_name(manifest)))
-                .collect::<Vec<_>>()
-                .join(", ")
-        } else {
-            cs.iter()
-                .map(|c| c.description(manifest))
+            let remove_msg = cs
+                .iter()
+                .map(|c| remove_component_msg(c, manifest, toolchain))
                 .collect::<Vec<_>>()
-                .join(", ")
-        };
+                .join("\n");
 
-        let remove_msg = cs
-            .iter()
-            .map(|c| remove_component_msg(c, manifest, toolchain))
-            .collect::<Vec<_>>()
-            .join("\n");
-        let _ = write!(
-            buf,
-            "some components unavailable for download for channel '{toolchain}': {cs_str}\n\
-            If you don't need the components, you can remove them with:\n\n{remove_msg}\n\n{TOOLSTATE_MSG}",
-        );
+            _ = write!(
+                buf,
+                "some components unavailable for download for channel '{toolchain}': {cs_str}"
+            );
+            if toolchain.starts_with("nightly") {
+                _ = write!(
+                    buf,
+                    "Sometimes not all components are available in any given nightly. "
+                );
+            }
+            _ = write!(
+                    buf,
+                    "If you are trying to update an existing toolchain and no longer need these components,\n\
+                    you might want to remove them first with:\n\n{remove_msg}\n\n{TOOLSTATE_MSG}",
+            );
+        }
```

Comment:
its quite hard to determine what you've actually changed here - the change combines a refactoring (fine), and new message  logic (also fine), but even though I've looked in:
- the PR message
- the commit messages
- test changes (none)
I can't tell what the actual intent is without manually stepping through in my head every possible case.

It would be a lot easier to reason about if the refactoring and the functional changes were separate (e.g. two commits in one PR - like you did for the other module that you refactored).

I realise that that isn't always possible, and if it wasn't here then I guess its as good as we can get :).

The other norm we have is to ensure test coverage for functional behaviours. We're in the process of moving tests closer to the domain for the test, so looking at the existing test coverage isn't necessarily the best way to figure out where tests should go - but its ok to put them in the blackbox test suite under `tests/` and we can move them later.

Reply:
> It would be a lot easier to reason about if the refactoring and the functional changes were separate (e.g. two commits in one PR - like you did for the other module that you refactored).

Yes, I am aware of that. I will leave the current commit as it is, but I will definitely improve this from now on.

> The other norm we have is to ensure test coverage for functional behaviours. We're in the process of moving tests closer to the domain for the test, so looking at the existing test coverage isn't necessarily the best way to figure out where tests should go - but its ok to put them in the blackbox test suite under `tests/` and we can move them later.

I'd be glad to add blackbox tests for this one (which will be done in another commit), please bear with me 🙏

Reply:
@rbtcollins The functional test changes now reflect this PR's original intention.